Vectorized: Use inline namespace instead of anonymous

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67656
* __->__ #67655
* #68555

Some of the CPU operators already use the `namespace CPU_CAPABILITY` trick to avoid anonymous namespacing, like [`PowKernel.cpp`](https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/aten/src/ATen/native/cpu/PowKernel.cpp#L14). This extends that pattern to the `Vectorized` class, which avoids `Wsubobject-linage` warnings like I was getting in #67621.

For many functions, it was necessary to add `inline` because the functions are defined in a header. There were no link errors previously because the anonymous namespace ensured they were not exposed to linkage. Similarly, free functions defined in an anonymous namespace might need the `C10_UNUSED` attribute to silence warnings about the function not being called in the only translation unit that it's defined in. By removing the anonymous namespace, these decorators are no longer necessary.

Differential Revision: [D32566109](https://our.internmc.facebook.com/intern/diff/D32566109)

peterbell10 committed:
Vectorized: Use inline namespace instead of anonymous

[ghstack-poisoned]

peterbell10 committed:
Update on "Vectorized: Use inline namespace instead of anonymous"




[ghstack-poisoned]

peterbell10 commented:
@malfet PTAL

malfet commented on aten/src/ATen/cpu/vec/vec256/vec256.h:
```diff
@@ -28,29 +28,28 @@
+// accessed as `at::vec`.
+inline namespace CPU_CAPABILITY {
+
+inline std::ostream& operator<<(std::ostream& stream, const c10::qint32& val) {
```
Can you please explain, why are you removing `C10_UNUSED` (I know that `inline` is sort of an equivalent, but `C10_UNUSED` has been validated across the plethora of compiler, so let's keep using it here)

peterbell10 replied:
When the function was static, each compilation unit has it's own copy and the compiler warns that it is unused; so this macro was needed to silence the warning. But it's not needed any more  because `inline` functions are expected to not be used in every single compilation unit they're defined.

malfet commented on aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:
```diff
@@ -581,32 +581,32 @@ Vectorized<BFloat16> inline operator^(const Vectorized<BFloat16>& a, const Vecto
 }
 
-Vectorized<BFloat16> Vectorized<BFloat16>::eq(const Vectorized<BFloat16>& other) const {
+inline Vectorized<BFloat16> Vectorized<BFloat16>::eq(const Vectorized<BFloat16>& other) const {
```
Please add the comment to PR description explaining why adding inline is necessary

malfet commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -843,25 +843,24 @@ inline mask_gather(const Vectorized<T>& src, T const* base_addr,
-    }
-  };
-}
+// There is a struct here because we don't have static_if and I can't
+// partially specialize a templated function.
+template<typename dst_t, typename src_t>
+struct CastImpl {
+  static inline Vectorized<dst_t> apply(const Vectorized<src_t>& src) {
+    src_t src_arr[Vectorized<src_t>::size()];
+    src.store(static_cast<void*>(src_arr));
+    return Vectorized<dst_t>::loadu(static_cast<const void*>(src_arr));
+  }
+};
+
+template<typename scalar_t>
+struct CastImpl<scalar_t, scalar_t> {
+  static inline Vectorized<scalar_t> apply(const Vectorized<scalar_t>& src) {
+    return src;
+  }
+};
```
Can you please move this change to a separate PR (as eliminating nested anonymous namespace is unrelated to this commit title)?

malfet reviewed (approved):
LGTM, but please address comments

peterbell10 committed:
Update on "Vectorized: Use inline namespace instead of anonymous"


Some of the CPU operators already use the `namespace CPU_CAPABILITY` trick to avoid anonymous namespacing, like [`PowKernel.cpp`](https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/aten/src/ATen/native/cpu/PowKernel.cpp#L14). This extends that pattern to the `Vectorized` class, which avoids `Wsubobject-linage` warnings like I was getting in #67621.


[ghstack-poisoned]

peterbell10 committed:
Update on "Vectorized: Use inline namespace instead of anonymous"


Some of the CPU operators already use the `namespace CPU_CAPABILITY` trick to avoid anonymous namespacing, like [`PowKernel.cpp`](https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/aten/src/ATen/native/cpu/PowKernel.cpp#L14). This extends that pattern to the `Vectorized` class, which avoids `Wsubobject-linage` warnings like I was getting in #67621.

For many functions, it was necessary to add `inline` because the functions are defined in a header but there were no link errors previously because the anonymous namespace ensured they were not exposed to linkage. Similarly, free functions defined in an anonymous namespace might need the `C10_UNUSED` attribute to silence warnings about the function not being called in the only translation unit that it's defined in. By removing the anonymous namespace, these decorators are no longer necessary.

[ghstack-poisoned]

peterbell10 commented:
@malfet I've moved the nested namespace removal into a new PR and added a note in this PR's description about `inline` and `C10_UNUSED` changes.

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32566109).

malfet commented:
Internal build complains about:
```
aten/src/ATen/native/cpu/PowKernel.cpp:14:11: error: inline namespace reopened as a non-inline namespace [-Werror]
namespace CPU_CAPABILITY {
```
Will propose the change to this PR as well as do corresponding internal change

malfet committed:
Make CPU_NAMESPACE inline in PowKernel

Make CPU_NAMESPACE inline here

malfet committed:
Update UnaryOpsKernel.cpp
